plutus-tx/src/PlutusTx/Bool.hs

```diff
@@ -38,6 +38,42 @@ infixr 2 ||
 not :: Bool -> Bool
 not a = if a then False else True
 
+{- Note [Short-circuit evaluation for `&&` and `||`]
+TLDR: as it stands, the behaviors of `&&` and `||` are inconsistent: they sometimes
+short-circuit, sometimes don't.
+
+The Plutus Tx compiler treats `if-then-else` and `case` specially, and ensures that
+the branches aren’t strictly evaluated. For example, `if b then exprT else exprF` is
+compiled to `(if b then (/\a -> exprT) (/\a -> exprF)) @(forall a. a)`. The type
```

Comment:
I've forgotten why we do this.  Is it because we can't use `()` because that doesn't necessarily exist in the current universe of built-in types and doing it with Scott units would introduce extra expense?  I was thinking that this seems like a bit of a hack and I don't recall having seen `/\` used like this before, but now I remember that we spent a long time (weeks/months) wondering whether `/\`should block evaluation or not and we eventually decided that it should.  I don't think there's consensus in the literature about whether evaluation should be allowed under unapplied `/\`.  [That doesn't have a lot to do with this PR: I'm just trying to refresh my memory about this stuff.]

Reply:
I don't remember exactly either; it could be that this is just easier to implement - all type abstractions and instantiations are simply compiled to `delay` and `force`. If we use `()` then it would have to be treated specially.

Reply:
> I've forgotten why we do this. Is it because we can't use `()` because that doesn't necessarily exist in the current universe of built-in types and doing it with Scott units would introduce extra expense?

`()` would require us to use a `\` which binds a useless name, which is less efficient than `/\` that compiles to a `delay`. I.e. basically we prefer `delay`-`force` to `lambda`-`apply`, since the former is both simpler and faster.

> `/\` should block evaluation or not and we eventually decided that it should.

I think that happened in [this](https://groups.google.com/a/iohk.io/g/plutus/c/GXPwl6o3xKE/m/HlxkO3hSBQAJ) Plutus mailing list thread.